src/main/java/org/skriptlang/skript/lang/experiment/Experiment.java

```diff
@@ -0,0 +1,181 @@
+/**
+ *   This file is part of Skript.
+ *
+ *  Skript is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Skript is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package org.skriptlang.skript.lang.experiment;
+
+import ch.njol.skript.patterns.PatternCompiler;
+import ch.njol.skript.patterns.SkriptPattern;
+import org.jetbrains.annotations.ApiStatus;
+
+import java.util.Objects;
+
+/**
+ * An optional, potentially-experimental feature enabled per-script with the {@code using X} syntax.
+ * Experiments provided by Skript itself are found in {@link Feature}.
+ * This can also represent an unknown experiment 'used' by a script that was not declared or registered
+ * by Skript or any of its addons.
+ */
+public interface Experiment {
+
+	@ApiStatus.Internal
+	static Experiment unknown(String text) {
+		return new UnmatchedExperiment(text);
+	}
+
+	/**
+	 * A constant experiment provider (designed for the use of addons).
+	 * @param codeName The debug 'code name' of this feature.
+	 * @param phase The stability of this feature.
+	 * @param patterns What the user may write to match the feature. Defaults to the codename if not set.
+	 * @return An experiment flag.
+	 */
+	static Experiment constant(String codeName, LifeCycle phase, String... patterns) {
+		return new ConstantExperiment(codeName, phase, patterns);
+	}
+
+	/**
+	 * @return The patterns that can be written in the `using` structure.
+	 */
+	default String[] patterns() {
+		return new String[] {this.codeName()};
+	}
+
+	/**
+	 * A simple, printable code-name for this pattern for warnings and debugging.
+	 * Ideally, this should be matched by one of the {@link #patterns()} entries.
+	 *
+	 * @return The code name of this experiment.
+	 */
+	String codeName();
+
+	/**
+	 * @return The safety phase of this feature.
+	 */
+	LifeCycle phase();
+
+	/**
+	 * @return Whether this feature was declared by Skript or a real extension.
+	 */
+	default boolean isKnown() {
+		return this.phase() != LifeCycle.UNKNOWN;
+	}
+
+	SkriptPattern compiledPattern();
+
+	default boolean matches(String text) {
+		return this.compiledPattern().match(text) != null;
+	}
+
+}
+
+/**
+ * A class for constant experiments.
+ */
+class ConstantExperiment implements Experiment {
```

Comment:
In some other places/PR, something like an `XImpl` is used. Do you think  a package-private class such as `ExperimentImpl` would be preferred? I'm not actually sure whether there is a meaningful difference.